Add back 8-zts-alpine, take 2

tianon opened the pull request:
#1075, but this time with proper gusto

Closes #1074

tianon commented:
blargh

WyriHaximus commented on 8.0-rc/alpine3.12/zts/Dockerfile:
```diff
@@ -0,0 +1,187 @@
+
+ENV GPG_KEYS 1729F83938DA44E27BA0F4D3DBDB397470D12172 BFDDD28642824F8118EF77909B67A5C12229118F
+
+ENV PHP_VERSION 8.0.0RC2
+ENV PHP_URL="https://downloads.php.net/~pollita/php-8.0.0RC2.tar.xz" PHP_ASC_URL="https://downloads.php.net/~pollita/php-8.0.0RC2.tar.xz.asc"
+ENV PHP_SHA256="97d8031ba1095c3222fc5725710ac18dc4f584e526cbcbfb2817146abb6891ec" PHP_MD5=""
```
Not that it helps for the error this PR is getting, but updated it to RC5 with no improvement

```suggestion
ENV PHP_VERSION 8.0.0RC5
ENV PHP_URL="https://downloads.php.net/~carusogabriel/php-8.0.0RC5.tar.xz" PHP_ASC_URL="https://downloads.php.net/~carusogabriel/php-8.0.0RC5.tar.xz.asc"
ENV PHP_SHA256="fa2e98d8b42e497a45f491c4714f0a67f7d817ff2f7ab9aad09efb72d2a8c60b" PHP_MD5=""
```

tianon replied:
Great point, thanks!  Updated (might as well keep it current in case things change). :+1:

WyriHaximus commented:
The error I keep coming back to is:

```
_tsrm_ls_cache: initial-exec TLS resolves to dynamic definition
```

Some Googling seems to suggest a musl issue, but I've also tried building with Alpine 3.11 and got the same error, as with Alpine 3.10.

Full error: 

```
Warning: PHP Startup: Unable to load dynamic library 'pdo_mysql.so' (tried: /usr/local/lib/php/extensions/no-debug-zts-20200930/pdo_mysql.so (Error relocating /usr/local/lib/php/extensions/no-debug-zts-2
0200930/pdo_mysql.so: _tsrm_ls_cache: initial-exec TLS resolves to dynamic definition in /usr/local/lib/php/extensions/no-debug-zts-20200930/pdo_mysql.so), /usr/local/lib/php/extensions/no-debug-zts-2020
0930/pdo_mysql.so.so (Error loading shared library /usr/local/lib/php/extensions/no-debug-zts-20200930/pdo_mysql.so.so: No such file or directory)) in Unknown on line 0
```

WyriHaximus commented:
Also just tried edge which comes with musl 1.2 (where 3.12 comes with 1.1) and still getting the same error. Getting the feeling, this issue purely comes from PHP 8 itself :(

WyriHaximus commented:
After reading https://wiki.php.net/zts-improvement and not having a clue what it means I also found https://bugs.freedesktop.org/show_bug.cgi?id=35268 which seems to suggest this needs to change in PHP itself. (As far as I understand.)

WyriHaximus commented:
@carusogabriel If you have any clue about this your help would be appreciated.

WyriHaximus commented:
And the issue is also present in 8.0.0 final

mvorisek commented:
Any progress on this PR so TS PHP needed for threaded apps is available?

tianon commented:
To be clear -- any actual progress on this PR is likely to happen in PHP upstream, not in the image (however, Alpine is not exactly well-supported, so I wouldn't suggest holding your breath).

I've rebased and added the 8.1 variants too so we can see whether anything has progressed.

mvorisek commented:
If anyone wants to help, here is extracted generated job source:

```
      {
        "name": "8.1.0alpha2-zts-alpine3.13",
        "os": "ubuntu-latest",
        "meta": {
          "entries": [
            {
              "name": "8.1.0alpha2-zts-alpine3.13",
              "tags": [
                "php:8.1.0alpha2-zts-alpine3.13",
                "php:8.1-rc-zts-alpine3.13"
              ],
              "directory": "8.1-rc/alpine3.13/zts",
              "file": "Dockerfile",
              "constraints": [],
              "froms": [
                "alpine:3.13"
              ]
            }
          ],
          "froms": [
            "alpine:3.13"
          ],
          "dockerfiles": [
            "8.1-rc/alpine3.13/zts/Dockerfile"
          ]
        },
        "runs": {
          "build": "docker build --tag 'php:8.1.0alpha2-zts-alpine3.13' --tag 'php:8.1-rc-zts-alpine3.13' '8.1-rc/alpine3.13/zts'",
          "history": "docker history 'php:8.1.0alpha2-zts-alpine3.13'",
          "test": "~/oi/test/run.sh 'php:8.1.0alpha2-zts-alpine3.13'",
          "prepare": "git clone --depth 1 https://github.com/docker-library/official-images.git -b master ~/oi\n# create a dummy empty image/layer so we can --filter since= later to get a meaningful image list\n{ echo FROM busybox:latest; echo RUN :; } | docker build --no-cache --tag image-list-marker -\n# PGP Happy Eyeballs\ngit clone --depth 1 https://github.com/tianon/pgp-happy-eyeballs.git ~/phe\n~/phe/hack-my-builds.sh\nrm -rf ~/phe",
          "pull": "docker pull 'alpine:3.13'",
          "images": "docker image ls --filter since=image-list-marker"
        }
      },
```

with this details, it should be possible to debug this issue localy.

mvorisek commented:
Is the issue still present /w Alpine latest/3.15?

tianon commented:
Thanks for the reminder!  I've just rebased so we can find out. :smile:

mvorisek commented:
I have opened https://github.com/php/php-src/issues/8160 in php-src. I do not know the issue directly and the CI output seems very limited, so if anyone knows the problem, please describe in the php-src issue, preferably with minimalistic Dockerfile to reproduce.

cmb69 commented:
Can you please try with the following patch:
````.diff
 ext/pdo/config.m4 | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ext/pdo/config.m4 b/ext/pdo/config.m4
index 9b9a3e36df..63a2f7ed0f 100644
--- a/ext/pdo/config.m4
+++ b/ext/pdo/config.m4
@@ -9,7 +9,7 @@ if test "$PHP_PDO" != "no"; then
   dnl Make sure $PHP_PDO is 'yes' when it's not 'no' :)
   PHP_PDO=yes
 
-  PHP_NEW_EXTENSION(pdo, pdo.c pdo_dbh.c pdo_stmt.c pdo_sql_parser.c pdo_sqlstate.c, $ext_shared)
+  PHP_NEW_EXTENSION(pdo, pdo.c pdo_dbh.c pdo_stmt.c pdo_sql_parser.c pdo_sqlstate.c, $ext_shared,,-DZEND_ENABLE_STATIC_TSRMLS_CACHE=1)
   PHP_ADD_EXTENSION_DEP(pdo, spl, true)
   PHP_INSTALL_HEADERS(ext/pdo, [php_pdo.h php_pdo_driver.h php_pdo_error.h])
 
````

tianon commented:
Here's the updated warning on 8.1 with that patch:

`Warning: Failed loading Zend extension 'opcache' (tried: /usr/local/lib/php/extensions/no-debug-zts-20210902/opcache (Error loading shared library /usr/local/lib/php/extensions/no-debug-zts-20210902/opcache: No such file or directory), /usr/local/lib/php/extensions/no-debug-zts-20210902/opcache.so (Error relocating /usr/local/lib/php/extensions/no-debug-zts-20210902/opcache.so: std_object_handlers: initial-exec TLS resolves to dynamic definition in /usr/local/lib/php/extensions/no-debug-zts-20210902/opcache.so)) in Unknown on line 0`

tianon commented:
Something similar (but on `pdo_mysql.so` instead of `opcache.so`) on 8.0:

`Warning: PHP Startup: Unable to load dynamic library 'pdo_mysql.so' (tried: /usr/local/lib/php/extensions/no-debug-zts-20200930/pdo_mysql.so (Error relocating /usr/local/lib/php/extensions/no-debug-zts-20200930/pdo_mysql.so: _tsrm_ls_cache: initial-exec TLS resolves to dynamic definition in /usr/local/lib/php/extensions/no-debug-zts-20200930/pdo_mysql.so), /usr/local/lib/php/extensions/no-debug-zts-20200930/pdo_mysql.so.so (Error loading shared library /usr/local/lib/php/extensions/no-debug-zts-20200930/pdo_mysql.so.so: No such file or directory)) in Unknown on line 0`

(Maybe that fix wasn't quite enough / in the right place for 8.0?)

tianon commented:
(I've also just pushed https://github.com/docker-library/official-images/commit/ce7303633c573ae71de9c9222766ec5feff3f581 which should make the _next_ run of these tests in CI actually show the error/warning correctly, which will be a lot more useful. :facepalm:)

cmb69 commented:
Thanks for checking, @tianon! I'm a bit surprised due to
https://github.com/php/php-src/blob/PHP-8.0/TSRM/TSRM.h#L150-L156. So for `__MUSL__`, `tls_model("initial-exec")` shouldn't be used at all.

cmb69 commented:
See also https://github.com/php/php-src/commit/804420b011308846f8427b274da7e28ab37534a6. Since @derickr committed this, maybe he has some ideas.

mvorisek commented:
@tianon please test patching PHP 8.0+ versions with

https://github.com/php/php-src/compare/PHP-8.0...mvorisek:abd95290272087285dc2827d0fa0fbe141db6ff7 (with both commits applied)

mvorisek commented:
> Looks like that still did the trick for 8.1, but not 8.0. 😅

weird, very weird, my build is passing even for PHP 8.0

see https://github.com/mvorisek/image-php/commit/29fdca7e8c350f616aa138708bc0dca0c1ba16b8

here is a test: https://github.com/atk4/data/runs/5456421886?check_suite_focus=true#step:4:5 , the CI confirms ZTS

here is the full patching code: https://github.com/mvorisek/image-php/commit/29fdca7e8c350f616aa138708bc0dca0c1ba16b8#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR55-R108 , it is based on `7.4/alpine3.15/zts/` source from this repo

the image from my build is published under `ghcr.io/mvorisek/image-php:8.0-zts-base`, can you run the tests on it?

mvorisek commented:
I build from source https://github.com/mvorisek/image-php/commit/29fdca7e8c350f616aa138708bc0dca0c1ba16b8#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR81

That can be the issue, as the patching in this PR does not change the `configure` file in any way.

tianon commented:
Ahhh good catch -- we build from the released source tarballs, so indeed, regenerating `configure` is not something we do (especially since it can often have negative effects depending on our version of `autoconf`).  I'll throw that into my hack here, since I verified that your build does indeed pass the offending test.

tianon commented:
There we go; much better. :smile:

mvorisek commented:
https://github.com/php/php-src/pull/8180 merged, so ZTS build should pass starting with upcoming PHP 8.0.18 and 8.1.5 releases.

mvorisek commented:
The releases are out, can you please rebase/retest?

WyriHaximus commented:
🟢 😱 🎉 !

mvorisek commented:
![image](https://user-images.githubusercontent.com/2228672/166676175-7b0e61b6-82d9-4cbf-aab3-40938b6a8a07.png)

tianon commented:
Oops, forgot to come back to this -- thanks for the reminder! :+1:
